src/main/java/org/hydev/mcpm/client/injector/PluginLoader.java

```diff
@@ -1,45 +1,135 @@
 package org.hydev.mcpm.client.injector;
 
+import com.google.common.reflect.TypeToken;
+import org.bukkit.Bukkit;
+import org.bukkit.command.Command;
+import org.bukkit.command.PluginCommand;
+import org.bukkit.command.SimpleCommandMap;
+import org.bukkit.event.Event;
+import org.bukkit.plugin.InvalidDescriptionException;
+import org.bukkit.plugin.InvalidPluginException;
+import org.bukkit.plugin.Plugin;
+import org.bukkit.plugin.RegisteredListener;
+import org.hydev.mcpm.utils.PluginJarFile;
+
 import java.io.File;
+import java.io.IOException;
+import java.net.URLClassLoader;
+import java.util.*;
+
+import static org.hydev.mcpm.utils.ReflectionUtils.getPrivateField;
+import static org.hydev.mcpm.utils.ReflectionUtils.setPrivateField;
 
 /**
  * TODO: Write a description for this class!
  *
  * @author Azalea (https://github.com/hykilpikonna)
  * @since 2022-09-27
  */
-public class PluginLoader
+public class PluginLoader implements LoadBoundary, UnloadBoundary, ReloadBoundary
 {
-    /**
-     * Dynamically load a local plugin through JVM reflections and classloader hacks
-     *
-     * @param jar Local jar file path
-     */
+    @Override
+    public void loadPlugin(String name) throws PluginNotFoundException
+    {
+        // 1. Find plugin file by name
+        var nf = new PluginNotFoundException(name);
+        var dir = new File("plugins");
+        if (!dir.isDirectory()) throw nf;
+        var file = Arrays.stream(Optional.ofNullable(dir.listFiles()).orElseThrow(() -> nf))
```

Comment:
This is fine, but note that a stack trace is assembled when the exception is constructed IIRC. It's better to create your exception on the line where you throw it so you have exact information about when/what went wrong.

Note how orElseThrow takes a lambda with no parameters.

Reply:
Fixed.